File: StreamingMinecraftClone/src/main/java/io/mc/physics/Physics.java
package io.mc.physics;

import org.joml.Vector3f;
import org.jspecify.annotations.NonNull;

import dev.dominion.ecs.api.Dominion;
import dev.dominion.ecs.api.Entity;
import io.mc.core.Transform;
import io.mc.renderer.Renderer;
import io.mc.renderer.Style;
import io.mc.renderer.Styles;
import io.mc.utils.JMath;
import io.mc.world.Block;
import io.mc.world.BlockFormat;
import io.mc.world.BlockMap;
import io.mc.world.ChunkManager;
import io.mc.world.World;

public final class Physics {
    private static Vector3f gravity = new Vector3f(0, 20.0f, 0.0f);
    private static Vector3f terminalVelocity = new Vector3f(50.0f, 50.0f, 50.0f);
    private static final float physicsUpdateRate = 1.0f / 120.0f;

    public static void init() {

    }

    public static void update(@NonNull Dominion registry) {
        float accumulatedDeltaTime = 0.0f;
        accumulatedDeltaTime += World.getDeltaTime();

        int numUpdates = 0;

        while (accumulatedDeltaTime >= physicsUpdateRate) {
            registry.findEntitiesWith(Transform.class, Rigidbody.class, BoxCollider.class).forEach(entity -> {
                Rigidbody rb = entity.comp2();
                Transform transform = entity.comp1();
                BoxCollider boxCollider = entity.comp3();

                transform.position.add(rb.velocity.mul(physicsUpdateRate));
                rb.velocity.add(rb.acceleration.mul(physicsUpdateRate));

                if (rb.useGravity) {
                    // rb.velocity.y -= Math.clamp(rb.velocity.y, -1 * terminalVelocity.y,
                    // terminalVelocity.y);
                }
            });
        }
    }

    public static RaycastStaticResult raycastStatic(@NonNull Vector3f origin, @NonNull Vector3f normalDirection,
            float maxDistance,
            boolean draw) {
        RaycastStaticResult result = new RaycastStaticResult();
        result.hit = false;

        if (normalDirection.equals(new Vector3f())) {
            return result;
        }

        Style red = Styles.defaultStyle;
        red.color = 0xE90101;
        if (draw) {
            Renderer.drawLine(origin, origin.add(normalDirection.mul(maxDistance)), red);
        }

        // NOTE: Thank God for this paper
        // http://www.cse.yorku.ca/~amana/research/grid.pdf which outlines what I'm
        // doing here
        Vector3f rayEnd = origin.add(normalDirection.mul(maxDistance));
        // Do some fancy math to figure out which voxel is the next voxel
        Vector3f blockCenter = origin.ceil();
        Vector3f step = normalDirection.absolute();
        // Max amount we can step in any direction of the ray, and remain in the voxel
        Vector3f blockCenterToOriginSign = (blockCenter.sub(origin)).normalize();
        Vector3f goodNormalDirection = new Vector3f(
                normalDirection.x == 0.0f ? 1e-10f * blockCenterToOriginSign.x : normalDirection.x,
                normalDirection.y == 0.0f ? 1e-10f * blockCenterToOriginSign.y : normalDirection.y,
                normalDirection.z == 0.0f ? 1e-10f * blockCenterToOriginSign.z : normalDirection.z);
        Vector3f tDelta = ((blockCenter.add(step)).sub(origin)).div(goodNormalDirection);
        // If any number is 0, then we max the delta so we don't get a false positive
        if (tDelta.x == 0.0f)
            tDelta.x = 1e10f;
        if (tDelta.y == 0.0f)
            tDelta.y = 1e10f;
        if (tDelta.z == 0.0f)
            tDelta.z = 1e10f;
        Vector3f tMax = tDelta;
        float minTValue;
        do {
            // TODO: This shouldn't have to be calculated every step
            tDelta = (blockCenter.sub(origin)).div(goodNormalDirection);
            tMax = tDelta;
            // minTValue = FLT_MAX;
            if (tMax.x < tMax.y) {
                if (tMax.x < tMax.z) {
                    blockCenter.x += step.x;
                    // Check if we actually hit the block
                    if (doRaycast(origin, normalDirection, maxDistance, draw, blockCenter, step, result)) {
                        return result;
                    }
                    // tMax.x += tDelta.x;
                    minTValue = tMax.x;
                } else {
                    blockCenter.z += step.z;
                    if (doRaycast(origin, normalDirection, maxDistance, draw, blockCenter, step, result)) {
                        return result;
                    }
                    // tMax.z += tDelta.z;
                    minTValue = tMax.z;
                }
            } else {
                if (tMax.y < tMax.z) {
                    blockCenter.y += step.y;
                    if (doRaycast(origin, normalDirection, maxDistance, draw, blockCenter, step, result)) {
                        return result;
                    }
                    // tMax.y += tDelta.y;
                    minTValue = tMax.y;
                } else {
                    blockCenter.z += step.z;
                    if (doRaycast(origin, normalDirection, maxDistance, draw, blockCenter, step, result)) {
                        return result;
                    }
                    // tMax.z += tDelta.z;
                    minTValue = tMax.z;
                }
            }
        } while (minTValue < maxDistance);

        return result;
    }

    private static boolean doRaycast(Vector3f origin, Vector3f normalDirection, float maxDistance, boolean draw,
            Vector3f blockCorner,
            Vector3f step, RaycastStaticResult out) {
        Vector3f blockCenter = blockCorner.min((new Vector3f(0.5f).mul(step)));
        if (draw) {
            Renderer.drawBox(blockCenter, new Vector3f(1.0f, 1.0f, 1.0f), Styles.defaultStyle);
        }

        int blockId = ChunkManager.getBlock(blockCenter).id;
        // BlockFormat block = BlockMap.getBlock(blockId);
        if (blockId != BlockMap.NULL_BLOCK.id && blockId != BlockMap.AIR_BLOCK.id) {
            BoxCollider currentBox = new BoxCollider();
            currentBox.offset = new Vector3f();
            currentBox.size = new Vector3f(1.0f, 1.0f, 1.0f);
            Transform currentTransform = new Transform();
            currentTransform.position = blockCenter;

            Block block = ChunkManager.getBlock(currentTransform.position);
            BlockFormat blockFormat = BlockMap.getBlock(block.id);

            if (blockFormat.isSolid) {
                Vector3f min = currentTransform.position.sub((currentBox.size.mul(0.5f)).add(currentBox.offset));
                Vector3f max = currentTransform.position.add(currentBox.size.mul(0.5f)).add(currentBox.offset);
                float t1 = (min.x - origin.x)
                        / (JMath.compare(normalDirection.x, 0.0f, Float.MIN_VALUE) ? 0.00001f : normalDirection.x);
                float t2 = (max.x - origin.x)
                        / (JMath.compare(normalDirection.x, 0.0f, Float.MIN_VALUE) ? 0.00001f : normalDirection.x);
                float t3 = (min.y - origin.y)
                        / (JMath.compare(normalDirection.y, 0.0f, Float.MIN_VALUE) ? 0.00001f : normalDirection.y);
                float t4 = (max.y - origin.y)
                        / (JMath.compare(normalDirection.y, 0.0f, Float.MIN_VALUE) ? 0.00001f : normalDirection.y);
                float t5 = (min.z - origin.z)
                        / (JMath.compare(normalDirection.z, 0.0f, Float.MIN_VALUE) ? 0.00001f : normalDirection.z);
                float t6 = (max.z - origin.z)
                        / (JMath.compare(normalDirection.z, 0.0f, Float.MIN_VALUE) ? 0.00001f : normalDirection.z);

                float tmin = Math.max(Math.max(Math.min(t1, t2), Math.min(t3, t4)), Math.min(t5, t6));
                float tmax = Math.min(Math.min(Math.max(t1, t2), Math.max(t3, t4)), Math.max(t5, t6));
                if (tmax < 0 || tmin > tmax) {
                    // No intersection
                    return false;
                }

                float depth = 0.0f;
                if (tmin < 0.0f) {
                    // The ray's origin is inside the AABB
                    depth = tmax;
                } else {
                    depth = tmin;
                }

                out.point = origin.add(normalDirection.mul(depth));
                out.hit = true;
                out.blockCenter = currentTransform.position.add(currentBox.offset);
                out.blockSize = currentBox.size;
                out.hitNormal = out.point.sub(out.blockCenter);
                float maxComponent = Math.max(Math.abs(out.hitNormal.x),
                        Math.max(Math.abs(out.hitNormal.y), Math.abs(out.hitNormal.z)));
                out.hitNormal = Math.abs(out.hitNormal.x) == maxComponent
                        ? new Vector3f(1, 0, 0).mul(Math.signum(out.hitNormal.x))
                        : Math.abs(out.hitNormal.y) == maxComponent
                                ? new Vector3f(0, 1, 0).mul(Math.signum(out.hitNormal.y))
                                : new Vector3f(0, 0, 1).mul(Math.signum(out.hitNormal.z));
                return true;
            }
        }

        return false;
    }

    private static void resolveStaticCollision(Entity entity, Rigidbody rb, Transform transform,
            BoxCollider boxCollider) {
        throw new UnsupportedOperationException();
    }

    private static CollisionManifold staticCollisionInformation(Rigidbody r1, BoxCollider b1, Transform t1,
            BoxCollider b2, Transform t2) {
        throw new UnsupportedOperationException();
    }

    private static float getDirection(CollisionFace face) {
        switch(face) {
            case BACK:
                return 1f;
            case BOTTOM:
                return 1f;
            case FRONT:
                return -1.0f;
            case LEFT:
                return 1.0f;
            case NONE:
                return 0.0f;
            case RIGHT:
                return -1.0f;
            case TOP:
                return -1.0f;
            default:
                return 0.0f;

        }
        
    }

    private static void getQuadrantResult(final Transform t1,final Transform t2,final BoxCollider b2Expanded, final CollisionManifold res) {
        throw new UnsupportedOperationException();
    }

    private static float peneratrationAmount() {
        throw new UnsupportedOperationException();
    }

    private static Interval getInterval() {
        throw new UnsupportedOperationException();
    }

}

File: rminecraft/src/physics.rs
use glm::{vec3, Vector3};

pub struct RaycastStaticResult {
    point: Vector3<f32>,
    block_center: Vector3<f32>,
    block_size: Vector3<f32>,
    hit_normal: Vector3<f32>,
    hit: bool,
}
impl RaycastStaticResult {
    fn default() -> RaycastStaticResult {
        return RaycastStaticResult {
            point: vec3(0.0, 0.0, 0.0),
            block_center: vec3(0.0, 0.0, 0.0),
            block_size: vec3(0.0, 0.0, 0.0),
            hit_normal: vec3(0.0, 0.0, 0.0),
            hit: false,
        };
    }
}

pub struct Rigidbody {
    velocity: Vector3<f32>,
    acceleration: Vector3<f32>,
    on_ground: bool,
    is_sensor: bool,
    use_gravity: bool,
}

impl Rigidbody {
    fn zero_forces(&mut self) {
        self.acceleration = vec3(0.0, 0.0, 0.0);
        self.velocity = vec3(0.0, 0.0, 0.0);
    }
}

pub struct BoxCollider {
    size: Vector3<f32>,
    offset: Vector3<f32>,
}

pub enum CollisionFace {
    NONE = 0,
    TOP = 1,
    BOTTOM = 2,
    BACK = 3,
    FRONT = 4,
    LEFT = 5,
    RIGHT = 6,
}

struct Interval {
    min: f32,
    max: f32,
}

impl Interval {
    fn default() -> Interval {
        Interval {min: 0.0,max:  0.0}
    }
}

struct CollisionManifold {
    overlap: Vector3<f32>,
    face: CollisionFace,
    did_collide: bool,
}

pub mod Physics {
    use std::ops::Add;
    use std::ops::Mul;

    use crate::ecs::Transform;

    use super::BoxCollider;
    use super::CollisionFace;
    use super::Interval;
    use super::RaycastStaticResult;
    use glm::vec3;
    use glm::ApproxEq;
    use glm::Vector3;

    const GRAVITY: Vector3<f32> = Vector3::<f32> {
        x: 0.0,
        y: 20.0,
        z: 0.0,
    };
    const TERMINAL_VELOCITY: Vector3<f32> = Vector3::<f32> {
        x: 50.0,
        y: 50.0,
        z: 50.0,
    };

    const PHYSICS_UPDATE_RATE: f32 = 1.0 / 120.0;

    fn init() {}

    fn update(registry: shipyard::World) {
        //    {static mut accumulated_delta_time: Lazy<f32> = Lazy::new(0.0);

        //     // accumulated_delta_time = 0.0;
        //     }
    }

    fn raycastStatic(
        origin: &Vector3<f32>,
        normal_direction: &Vector3<f32>,
        max_distance: f32,
        draw: bool,
        block_corner: &Vector3<f32>,
        step: Vector3<f32>,
    ) -> RaycastStaticResult {
        let mut result: RaycastStaticResult = RaycastStaticResult::default();

        result.hit = false;

        if (normal_direction.is_approx_eq(&vec3(0.0, 0.0, 0.0))) {
            return result;
        }

        if (draw) {}

        let ray_end = origin.add(normal_direction.mul(max_distance));

        todo!()
    }

    fn resolve_static_collision() {}

    fn static_collision_information() {}

    fn get_direction(face: &CollisionFace) -> f32 {
        match face {
            CollisionFace::NONE => return -1.0,
            CollisionFace::TOP => todo!(),
            CollisionFace::BOTTOM => todo!(),
            CollisionFace::BACK => todo!(),
            CollisionFace::FRONT => todo!(),
            CollisionFace::LEFT => todo!(),
            CollisionFace::RIGHT => todo!(),
        }
    }

    fn get_quadrant_result() {}

    fn is_colliding() {}

    fn penetration_amount() {}

    fn get_interval(
        box_collider: &BoxCollider,
        transform: &Transform,
        axis: &Vector3<f32>,
    ) -> Interval {
        let mut vertices = Vec::new();

        let center = transform.position;
        let half_size = box_collider.size * 0.5;

        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z) + vec3(0.0, 1.0, 0.0) * (half_size.y)
                - vec3(1.0, 0.0, 0.0) * (half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));
        vertices.push(center.add(
            vec3(0.0, 0.0, 1.0) * (half_size.z)
                + vec3(0.0, 1.0, 0.0).mul(half_size.y)
                + vec3(1.0, 0.0, 0.0).mul(half_size.x),
        ));

        let result = Interval::default();
        // let dot_product = axis.dot( &vertices[0]);

        for vertex in vertices {
            
        }

        todo!()
    }

    fn close_point_on_ray() {}
}
